javalib/src/main/scala/java/net/Inet6Address.scala

```diff
@@ -1,27 +1,56 @@
 package java.net
 
-// Ported from Apache Harmony
+// Ported from Apache Harmony & extensively re-worked for Scala Native.
 
 import scalanative.unsafe._
 import scalanative.unsigned._
 
+import java.{util => ju}
+
 import scala.scalanative.posix.net.`if`._
 import scala.scalanative.posix.stddef
 
 final class Inet6Address private (
     ipAddress: Array[Byte],
     host: String,
+    useScopeId: Boolean, // true when this created with explicit, valid scopeId
     scopeId: Int,
-    val zoneIdent: String
+    nif: NetworkInterface
 ) extends InetAddress(ipAddress, host) {
 
-  private[net] def this(ipAddress: Array[Byte], host: String, scope: Int) =
-    this(ipAddress, host, scope, "")
+  /* Note on scopeId constructor argument:
+   *
+   *   For most intents and purposes a Java scopeId is a NetworkInterface index
+   *   equivalent to a C or IETF (Internet Engineering Task Force)
+   *   sin6_scope_id.
+   *
+   *   IETF discussions sometimes call it a 'zoneId'.
+   *
+   *   See the "Design Note_2" in NetworkInterface.scala for a description of
+   *   a case where Java and C practice differ.
+   */
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Inet6Address => this.hashCode() == that.hashCode()
+    case _                  => false
+  }
+
+  def getScopedInterface(): NetworkInterface = nif
 
-  private[net] def this(ipAddress: Array[Byte], host: String) =
-    this(ipAddress, host, 0)
+  def getScopeId(): Int =
+    if (useScopeId) scopeId
+    else if (nif == null) 0
+    else nif.getIndex()
 
-  def getScopeId(): Int = scopeId
+  override def hashCode(): Int = {
+    val base = ju.Arrays.hashCode(ipAddress)
+      + host.hashCode()
+      + useScopeId.hashCode()
+      + scopeId.hashCode()
```

Comment:
A good practice is to combine hashes by first multiply accumulated hashCode with some prime number to reduce chance of conflicts, eg. ` 
```
var res: Int = 1
res = 31 * res + ju.Arrays.hashCode(ipAddress) 
res = 31 * res + host.hashCode()
res = 31 * res + useScopeId.hashCode()
res = 31 * res + scopeId.hashCode()
res
```
see, `java.util.Arrays.hashCode` or  our implementation of it in `java.util.Arrays.hashCodeImpl`

Reply:
Wojciech,

Thank you for the review and, especially, the welcome short tutorial. I always learn things.

I reviewed the other uses of hashcode in `java.net` and will be updating those usages.